From 51d2d6a9cf70c7e85b15456527fa06b89848b720 Mon Sep 17 00:00:00 2001
From: David Milburn <dmilburn@redhat.com>
Date: Mon, 5 May 2014 21:04:06 -0400
Subject: [block] nvme: Fix divide-by-zero in nvme_trans_io_get_num_cmds

Message-id: <1399323848-28227-11-git-send-email-dmilburn@redhat.com>
Patchwork-id: 79400
O-Subject: [RHEL7.1 PATCH BZ 1081734 10/12] NVMe: Fix divide-by-zero in nvme_trans_io_get_num_cmds
Bugzilla: 1081734
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: Keith Busch <keith.busch@intel.com>

dev->max_hw_sectors may be zero to indicate the device has no limit on
the number of sectors.  nvme_trans_do_nvme_io() should use the software
limit, since this is guaranteed to be non-zero.

Reported-by: Mundu <mundu2510@gmail.com>
Signed-off-by: Keith Busch <keith.busch@intel.com>
Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
(cherry picked from commit ddcb776286c091189a7b928188112470ec7e9efc)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/nvme-scsi.c b/drivers/block/nvme-scsi.c
index e157e85..111c920 100644
--- a/drivers/block/nvme-scsi.c
+++ b/drivers/block/nvme-scsi.c
@@ -2044,7 +2044,7 @@ static int nvme_trans_do_nvme_io(struct nvme_ns *ns, struct sg_io_hdr *hdr,
  struct nvme_command c;
  u8 opcode = (is_write ? nvme_cmd_write : nvme_cmd_read);
  u16 control;
- u32 max_blocks = nvme_block_nr(ns, dev->max_hw_sectors);
+ u32 max_blocks = queue_max_hw_sectors(ns->queue);
 
  num_cmds = nvme_trans_io_get_num_cmds(hdr, cdb_info, max_blocks);
 
-- 
1.7.1